### ISSUE TITLE
Custom datasets and inference

### ISSUE THREAD

**Issue by jruiz-aplitop** [prs-eth/LitePT#23]
Hello.

Thanks for such an amazing project and paper, its very visual and well explained. 

I was wondering if there's a plan to add support for custom datasets or even multiple datasets (like a combination of S3DIS and Scannet) to see how well does the model generalizes and behaves with different data distributions (following Appendix C.2)

Thanks again.

Bests,
Javier.

**Comment by ywyue** [comment]
Hi @jruiz-aplitop, sorry, I just saw your issue now. Supporting custom datasets or multiple datasets is straightforward.
- For custom datasets, just organize your datasets as follows:
```
Data/
└──Your_Dataset_Name /
     └──Train /
         └── scene1_name /
               ├── coord.npy
               ├── segment.npy # if applicable, semantic label,
               ├── color.npy # if applicable, RGB
               ├── normal.npy # if applicable, normal
               └── any other attributes
         └── scene2_name
         └── ...
     └──val / # if applicable, validation set
     └──test / # if applicable, test set
```
Then, you can use `dataset_type = "DefaultDataset"` in the config file.

- For multiple datasets joint training, you can use "ConcatDataset" in the config file. For example, let's say training semantic segmentation jointly on NuScenes and Waymo, define the training set as:
```
    train=dict(
        type="ConcatDataset",
        datasets=[
            # nuScenes
            dict(
                type="NuScenesDataset",
                split="train",
                data_root="data/nuscenes",
                transform=data_transform_nuscenes,
                test_mode=False,
                ignore_index=-1,
                loop=1,
            ),
            # Waymo
            dict(
                type="WaymoDataset",
                split="training",
                data_root="data/waymo",
                transform=data_transform_waymo,
                test_mode=False,
                ignore_index=-1,
                loop=1,
            ),
        ],
    ),
```

But you need to make sure the semantic categories of multiple datasets are aligned, i.e. same semantic classes.

**Comment by jruiz-aplitop** [comment]
Great, thanks a lot @ywyue .

My plan is to try what you just describe also, multiple datasets joint training, for both outdoor and indoor scenes, to see how this model behaves.

What parameters do you think are more important to tweak between use cases? (apart from obviously the voxel size).

Thanks again! 

Pd. Just wanted to say congratulations again, not many papers in this 3D area are as visual and well explained as this one.

Bests,
Javier.

**Comment by ywyue** [comment]
Hi @jruiz-aplitop, thanks a lot for your kind words. 

> multiple datasets joint training, for both outdoor and indoor scenes

Do you mean training a single model across outdoor and indoor scenes? 

If so, I think attention needs to be paid to the following aspects:
- Semantic label alignment as I mentioned before.
- Voxel size as you mentioned.
- Data augmentation.
- Input features `(feat_keys):` all datasets should have xyz for sure; indoor datasets usually have color and normal, while outdoor datasets usually have strength. You may consider selectively use them whenever available.
- Dataset sampling balance: dataset sizes can differ substantially. With `ConcatDataset`, consider adjust the `loop`, which controls the sampling ratio so that one dataset does not dominate optimization.
- Cropping / number of input points: indoor and outdoor, or different datasets have different point density, counts, spatial extents, so parameters such as crop size or `point_max` deserve attention.

**Comment by jruiz-aplitop** [comment]
Hi @ywyue , thanks a lot for your response.

I was mainly talking about mixing datasets with different data distributions (like for example, different sensor or data acquisition methods, like LIDAR, SLAM, photogrametry, ...) into a same model, but one for outdoor and a different one for indoor, as voxel size is a main constraint for precision / efficiency between these two and is probably not the best idea to mix them.

I assume the same aspects you presented also apply to this, but if you feel like anything else should be mentioned, please, do so.

What I am looking for is to see how well these models can generalize, maybe not in terms of labels, but in terms of adapting to any kind of data acquisition and scene (keeping, as said before, a distinction between outdoor and indoor scenes) and using general labels, like ground, building, vertical element, vehicle, vegetation, ... to avoid problems with labels.

Thanks a lot again for everything.
Bests,
Javier.

**Comment by ywyue** [comment]
Hi @jruiz-aplitop, one point that might be helpful in your case is to use LayerNorm throughout the network instead of BatchNorm. This may improve generalization when joint training on multiple datasets.

**Comment by jruiz-aplitop** [comment]
Hi @ywyue , thanks again for your response.
Do you mean literally replacing all BatchNorm layer of the network to LayerNorm? Or using the built-in dataset-aware normalization setup, like shown below?:

```
backbone=dict(
    #  ...
    pdnorm_bn=True,
    pdnorm_ln=True,
    pdnorm_decouple=True,
    pdnorm_adaptive=False,
    pdnorm_affine=True,
    pdnorm_conditions=("DatasetA", "DatasetB", "DatasetC"),
) 
```
Thanks!

**Comment by ywyue** [comment]
Hi @jruiz-aplitop, I meant replacing all BatchNorm layer of the network to LayerNorm. I'd prefer to avoid dataset-aware normalization, as it requires dataset-specific handling and less scalable. But feel free to try both approaches depending on your needs.
